internal/memhttp/memhttp.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2021-2023 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memhttp
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/h2c"
+)
+
+// Server is a net/http server that uses in-memory pipes instead of TCP. By
+// default, it supports http/2 via h2c. It otherwise uses the same configuration
+// as the zero value of [http.Server].
+type Server struct {
+	server         http.Server
+	listener       *MemoryListener
+	url            string
+	cleanupTimeout time.Duration
+	disableHTTP2   bool
+
+	serverWG  sync.WaitGroup
+	serverErr error
+}
+
+// NewServer creates a new Server that uses the given handler. Configuration
+// options may be provided via [Option]s.
+func NewServer(handler http.Handler, opts ...Option) *Server {
+	var cfg config
+	WithCleanupTimeout(5 * time.Second).apply(&cfg)
+	for _, opt := range opts {
+		opt.apply(&cfg)
+	}
+
+	if !cfg.DisableHTTP2 {
+		h2s := &http2.Server{}
+		handler = h2c.NewHandler(handler, h2s)
+	}
+	listener := NewMemoryListener()
+	server := &Server{
+		server: http.Server{
+			Handler:           handler,
+			ReadHeaderTimeout: 5 * time.Second,
+		},
+		listener:       listener,
+		cleanupTimeout: cfg.CleanupTimeout,
+		url:            "http://" + listener.Addr().String(),
+		disableHTTP2:   cfg.DisableHTTP2,
+	}
+	server.goServe()
+	return server
+}
+
+// Transport returns a [http.RoundTripper] configured to use in-memory pipes
+// rather than TCP and talk HTTP/2 (if the server supports it).
+func (s *Server) Transport() http.RoundTripper {
+	if s.disableHTTP2 {
```

Comment:
I think this should always return a transport that uses HTTP/2.

To test things that require HTTP 1.1 (like tests in this package to verify that bidi requests gracefully fail with 505 status), we can also have a `TransportHTTP1` method that returns an HTTP/1.1-only transport. I don't think we need a `ClientHTTP1` because a caller can always do this instead:
```go
client := svr.Client()
client.Transport = svr.TransportHTTP1()
```

We'd likely also need this (if/when we export this) for callers that want to do stuff with the server like web sockets (which isn't yet supported on HTTP/2 in the Go runtime or 3rd party web sockets packages).